src/symbols.c

```diff
@@ -1914,27 +1867,276 @@ static void load_user_tags(filetype_id ft_id)
+	gtk_window_set_destroy_with_parent(GTK_WINDOW(tag_goto_popup), TRUE);
+	gtk_window_set_position(GTK_WINDOW(tag_goto_popup), GTK_WIN_POS_CENTER_ON_PARENT);
+	gtk_widget_set_size_request(tag_goto_popup, 250, 150);
+
+	frame = gtk_frame_new(NULL);
+	gtk_container_add(GTK_CONTAINER(tag_goto_popup), frame);
+	gtk_frame_set_shadow_type(GTK_FRAME(frame), GTK_SHADOW_OUT);
+	gtk_container_set_border_width(GTK_CONTAINER(frame), 0);
+
+	scroller = gtk_scrolled_window_new(NULL, NULL);
+	gtk_container_set_border_width(GTK_CONTAINER(scroller), 0);
+	gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(scroller),
+		GTK_POLICY_NEVER, GTK_POLICY_AUTOMATIC);
+	gtk_container_add(GTK_CONTAINER(frame), scroller);
+
+	/* TreeView and its model */
+	store = gtk_list_store_new(N_COLUMNS, GDK_TYPE_PIXBUF, G_TYPE_STRING, TM_TYPE_TAG);
+	tag_goto_tree_view = gtk_tree_view_new_with_model(GTK_TREE_MODEL(store));
+	gtk_widget_set_can_focus(tag_goto_tree_view, TRUE);
+	gtk_container_add(GTK_CONTAINER(scroller), tag_goto_tree_view);
+	g_signal_connect(tag_goto_tree_view, "focus-out-event", G_CALLBACK(on_focus_out), tag_goto_popup);
+	g_signal_connect(tag_goto_tree_view, "row-activated", G_CALLBACK(on_row_activated), NULL);
+	g_signal_connect(tag_goto_tree_view, "key-press-event", G_CALLBACK(on_key_pressed), NULL);
+
+	selection = gtk_tree_view_get_selection(GTK_TREE_VIEW(tag_goto_tree_view));
+	gtk_tree_selection_set_mode(selection, GTK_SELECTION_SINGLE);
+	gtk_tree_view_set_headers_visible(GTK_TREE_VIEW(tag_goto_tree_view), FALSE);
+	gtk_tree_view_set_reorderable(GTK_TREE_VIEW(tag_goto_tree_view), FALSE);
+
+	/* Columns */
+	column = gtk_tree_view_column_new();
+	gtk_tree_view_column_set_sizing(column, GTK_TREE_VIEW_COLUMN_FIXED);
+
+	renderer = gtk_cell_renderer_pixbuf_new();
+	gtk_tree_view_column_pack_start(column, renderer, FALSE);
+	gtk_tree_view_column_add_attribute(column, renderer, "pixbuf", PIXBUF_COLUMN);
+
+	renderer = gtk_cell_renderer_text_new();
+	gtk_tree_view_column_pack_start(column, renderer, TRUE);
+	gtk_tree_view_column_add_attribute(column, renderer, "markup", TEXT_COLUMN);
+
+	gtk_tree_view_append_column(GTK_TREE_VIEW(tag_goto_tree_view), column);
+
+	gtk_widget_show_all(frame);
+}
+
+
+static void show_goto_popup(GPtrArray *tags, gboolean have_best)
+{
+	gboolean first = TRUE;
+	GtkTreePath *path;
+	GtkTreeModel *model;
+	GtkTreeIter iter;
+	TMTag *tmtag;
+	gint i;
+
+	if (!tag_goto_popup)
+		create_goto_popup();
+
+	model = gtk_tree_view_get_model(GTK_TREE_VIEW(tag_goto_tree_view));
+	gtk_list_store_clear(GTK_LIST_STORE(model));
+
+	foreach_ptr_array(tmtag, i, tags)
+	{
+		gchar *fname = g_path_get_basename(tmtag->file->file_name);
+		gchar *text;
+
+		if (first && have_best)
+			text = g_strdup_printf("<b>%s: %lu</b>", fname, tmtag->line);
+		else
+			text = g_strdup_printf("%s: %lu", fname, tmtag->line);
+
+		gtk_list_store_insert_with_values(GTK_LIST_STORE(model), &iter, -1,
+				PIXBUF_COLUMN, get_tag_icon(get_tag_class(tmtag)),
```

Comment:
I suggest something like this https://github.com/b4n/geany/commit/5533a5031780867a97e157d200dba8c6f8bfb569